File: android/app/src/main/java/br/com/assistatecnologia/uaa_analyser/Analyser.java
package br.com.assistatecnologia.uaa_analyser;

import android.util.Log;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;

public class Analyser
{
    private static final long MAX_DELAY = 2750;

    private Message lastMessage = null;
    private int messageCount = 0;
    private List<Interval> longDelays = new ArrayList<>();
    private LinkedList<Integer> lostMessages = new LinkedList<>();
    private List<Integer> unorderedMessages = new ArrayList<>();
    private int delaySum = 0;
    private int delayCount = 0;

    public int getLocalCount() {
        return this.messageCount;
    }

    public float getAverageDelay()
    {
        if (this.delayCount == 0) return 0;
        return this.delaySum / this.delayCount;
    }

    public List<Interval> getLongDelays() {
        return Collections.synchronizedList(this.longDelays);
    }

    public int getLostMessagesCount() {
        return this.lostMessages.size();
    }

    public int getUnorderedMessagesCount() {
        return this.unorderedMessages.size();
    }

    public void add(Message message)
    {
        if (this.messageCount > 0)
        {
            if (message.count < this.lastMessage.count)
            {
                Log.d("Analyser", String.format("Received unordered message (%s)",
                        message.count));

                this.unorderedMessages.add(message.count);
                this.lostMessages.remove(message.count);
            }
            else
            {
                long delay = message.time - this.lastMessage.time;

                if (delay > MAX_DELAY)
                    this.longDelays.add(new Interval(this.lastMessage.time, message.time));
                else {
                    this.delayCount++;
                    this.delaySum += delay;
                }

                int lost = this.lastMessage.count + 1;
                while (lost < message.count) this.lostMessages.add(lost++);
            }
        }

        this.lastMessage = message;
        this.messageCount++;
    }
}

File: node/src/analyser.js
const MulticastClient = require('./multicast-client');
const config = require('./config');
const fs = require('fs');
const path = require('path');


const MAX_DELAY = 2750;

class Message {
  constructor (message) {
    this.time = new Date().getTime();
    this.counter = parseInt(message.counter);
  }
}

class Interval {
  constructor (begin, end) {
    this.begin = begin;
    this.end = end;
  }

  get time () {
    return this.end - this.begin;
  }
}

class Analyser {
  constructor (saved)
  {
    this.lastMessage = saved ? saved.lastMessage : null;
    this.messageCount = saved ? saved.messageCount : 0;;
    this.longDelayIntervals = saved ? saved.longDelayIntervals : [];
    this.lostMessages = saved ? saved.lostMessages : [];
    this.unorderedMessages = saved ? saved.unorderedMessages : [];
    
    this._delaySum = saved ? saved._delaySum : 0;
    this._delayCount = saved ? saved._delayCount : 0;
  }

  get averageDelay ()
  {
    if (this._delayCount === 0) return 0;
    return this._delaySum / this._delayCount;
  }

  addMessage (message)
  {
    if (this.messageCount > 0)
    {
      if (message.counter < this.lastMessage.counter)
      {
        console.log('Received unordered message (%s)', message.counter);
        this.unorderedMessages.push(message.count);

        const unorderedIndex = this.lostMessages.indexOf(message.counter);
        if (unorderedIndex >= 0) this.lostMessages.splice(unorderedIndex, 1);
      }
      else
      {
        const delay = message.time - this.lastMessage.time;
  
        if (delay > MAX_DELAY)
          this.longDelayIntervals.push(new Interval(this.lastMessage.time, message.time))
        else {
          this._delaySum += delay;
          this._delayCount += 1;
        }
  
        let lost = this.lastMessage.counter + 1;
        while (lost < message.counter) this.lostMessages.push(lost++);
      }
    }

    this.lastMessage = message;
    this.messageCount++;
  }
}


exports.Analyser = Analyser;
exports.Message = Message;
exports.Interval = Interval;